### DISCUSSION TITLE
Why editing content is harder than inserting new content in a PDF

### DISCUSSION THREAD

**Discussion by roylo97** [pymupdf/PyMuPDF#2685] [Looking for help] [answered]
I am trying to edit the content in existing PDF
but I am facing the difficult issues.

**### 1. Replacing existing texts or injecting texts in existing sentense**
Sometime, It was not what I expected. Such as missing font style, font size, and position issues (even if I copy the old properties)
```python
using these method

add_redact_annot()
apply_redactions()
``` 

Q: I am not sure if that is because the char of the font is not embedded in the PDF(the saved PDF compressed the font, so it will remove any useless char).

**### 2. Replacing image**
If I remove the old one and paste the new image,
there is a transparent frame in the old one, I saw it in Acrobat
https://pymupdf.readthedocs.io/en/latest/page.html#Page.delete_image

Q: Is there any way to remove it completely? or am I missing some methods to run after delete_image?

**### 3. Relocating text and image**
I am capable of adding new text in various locations, but repositioning it with the current elements is challenging. I came across a discussion on this topic in an old post at https://github.com/pymupdf/PyMuPDF/discussions/942. It appears that accomplishing this task may require utilizing a low-level method and performing updates, which can be quite challenging.

Q: is it still hard to edit the text position nowadays? 
Why we can easily insert the new text but hard to edit existing content? 
Is it because of some limitation or some format issue? 
I am not sure the online PDF editing tool seems easy to handle this (may be misleading that they also create a new one instead of editing the source)

**### 4. Adding custom Metadata in PDF**
```python
set_metadata({"producer":"test","demo":"test"})


it shows:
bad dict key(s): {'demo'}
``` 

Q: is it the limitation of the standard PDF format?

**### 5. About the comment in binary format**

Q: is it possible to add some comment/heading in binary format?
% xxxxxxx
https://community.adobe.com/t5/acrobat-discussions/what-does-is-this-binary-code-mean-%C3%A2%C3%A3%C3%AF%C3%B3-missing-from-some-files/m-p/9492857

**Comment by JorjMcKie** [comment]
I understand your frustration. What you are trying to do is ... sorry ... impossible.
Or possible only in the most simple constellations.
The reasons for this are complex if looking at details. But with a bird's eye view one may say: PDF is a "write-only" format, it is not **_meant_** to be edited.

All your issues go back to this:
* Subset fonts only contain the characters used in the PDF. New / replacing text in "that" font requires to locate the original (full) font somehow and use this instead. Plus you have to deal with spacing issues: new text usually is shorter or longer, plus you have to find the correct insertion point.
* the PDF creator may have used dirty tricks to simulate effects for text shading or have **_simulated_** bold or italics.
* Vector graphics (lines, rectangles, curves with complex rules to color areas) cannot be changed at all: they must be redrawn completely.
* Images are **_not_** always identified by xrefs: there exist inline images which only live inside a page's `/Contents` source. Inline images pose the same problem as vector graphics.

The above has led to converters: PDF -> Word for example. This forgets the PDF, then uses an office product to work with the result, then convert back to PDF. In the end, this is the only viable solution.
And even here, you will encounter frustrating situations: Tables in the PDF will often not result in tables in the word document, etc. etc.

Other questions:

* Custom PDF metadata are no problem, see documentation: https://pymupdf.readthedocs.io/en/latest/recipes-low-level-interfaces.html#how-to-extend-pdf-metadata
* Images, text and hyperlinks can be removed using redaction annotations: https://pymupdf.readthedocs.io/en/latest/page.html#Page.apply_redactions and https://pymupdf.readthedocs.io/en/latest/page.html#Page.add_redact_annot.
* Replacing / removing "xref-ed" images "globally" only works by simulating this using a fake microscopic transparent image instead. There is no other way to achieve this - except using redactions.

**Reply by roylo97** [reply]
Thanks for your explanation. 
Yes, It's hard to learn in PDF structure. 
The first time to saw the source code, I had no idea what Td, Tc Tj, are, etc....

**### Last question about comments in low level or binary etc...**

I would like to insert some comment messages in the near line(page.read_contents().splitlines()) / stream etc. 
To record which one is modified by my tool, not just the whole file(It should be mentioned in metadata), to trace the file record. 

Is it possible to add this?

**Reply by JorjMcKie** [reply]
You can add comment in the appearance code: use the % sign, the rest of the line will be ignored.
Note however, that cleaning actions like `page.clean_contents()` and certain save options will remove these comments.
You could use an `EmbeddedFiles` item to store any kind of information.

**Reply by roylo97** [reply]
Could you guide me or a few lines of code how to add comment in the appearance code(use the % sign)?

**Reply by JorjMcKie** [reply, marked as answer]
A page can have multiple `/Contents` objects. So the definition in the page object either looks something like `/Contents 7 0 R` or `/Contents [7 0 R 8 0 R ...]`. If you execute `page.clean_contents()`, multiple objects will be conctenated, the result replacing the previous lot.

So much for some background. `page.clean_contents()` does very much more than this concatenation: it also cleans and standardizes the syntax plus it makes sure that all resources used in the contents are in 1:1 correspondence with the objects named in the page's definition under `/Resources`. This applies to images, fonts and several more.

I recommend you first execute `page.clean_contents()`. Then `page.get_contents()` will deliver a list with exactly one item: the xref of the resulting contents stream.
In that stream (a `bytes` object), each command occurs in its own line. Your comment could either be appended to any of those lines, or you could just insert a new line. Example:
```python
xref = page.get_contents()[0]  # xref of /Contents stream
contlines = doc.xref_stream(xref).splitlines()  # contents stream decompressed and split into lines
contlines.insert(15, b"% comment before line 15")
line20 = contlines[20]
line20 += b" % comment appended to line 20"
contlines[20] = line20
# when finished:
doc.update_stream(xref, b"\n".join(contlines))
doc.ez_save("output.pdf")  # with garbage option
```

**Reply by roylo97** [reply]
Thanks @JorjMcKie 
I have another questions but I will create a new discussion.
